sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -423,31 +424,37 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
         // Otherwise, interpolate the number of partitions we need to try, but overestimate
         // it by 50%. We also cap the estimation in the end.
         val limitScaleUpFactor = Math.max(sqlContext.conf.limitScaleUpFactor, 2)
-        if (buf.isEmpty) {
+        if (scannedRowCount == 0) {
           numPartsToTry = partsScanned * limitScaleUpFactor
         } else {
-          val left = n - buf.size
+          val left = n - scannedRowCount
           // As left > 0, numPartsToTry is always >= 1
-          numPartsToTry = Math.ceil(1.5 * left * partsScanned / buf.size).toInt
+          numPartsToTry = Math.ceil(1.5 * left * partsScanned / scannedRowCount).toInt
           numPartsToTry = Math.min(numPartsToTry, partsScanned * limitScaleUpFactor)
         }
       }
 
       val p = partsScanned.until(math.min(partsScanned + numPartsToTry, totalParts).toInt)
       val sc = sqlContext.sparkContext
-      val res = sc.runJob(childRDD,
-        (it: Iterator[Array[Byte]]) => if (it.hasNext) it.next() else Array.empty[Byte], p)
-
-      buf ++= res.flatMap(decodeUnsafeRows)
-
+      val res = sc.runJob(childRDD, (it: Iterator[(Long, Array[Byte])]) =>
+        if (it.hasNext) it.next() else (0L, Array.empty[Byte]), p)
+
+      var i = 0
+      while (scannedRowCount < n && i < res.length) {
+        val rows = decodeUnsafeRows(res(i)._2)
```

Comment:
nit:
```
val rows = decodeUnsafeRows(res(i)._2)
val rowsToTake = if (n - scannedRowCount >= res(i)._1) {
  rows.toArray
} else {
  rows.take(n - scannedRowCount).toArray
}
scannedRowCount += rowsToTake.length
buf ++= rowsToTake
```

BTW do we still need `scannedRowCount`? It's always same as `buf.length`

Reply:
Thank you for review. Looks much better!

Reply:
Yes, we can remove scannedRowCount.